NitroxClient/GameLogic/Crafting.cs

```diff
@@ -9,27 +12,453 @@ namespace NitroxClient.GameLogic
 {
     public class Crafting
     {
+        internal static Crafting Instance; 
+        private System.Collections.Generic.Dictionary<string, GameLogic.CraftingUsageInfo> fabricatorsInUseByRemotePlayers = new System.Collections.Generic.Dictionary<string, GameLogic.CraftingUsageInfo>();
+        private readonly System.Timers.Timer timer = new System.Timers.Timer(5000);
+        private readonly int remoteSignalTimeOut = 120; //in seconds. 2 Minutes, because if someone crafts multiple items, no new starttime is sent and we don't want to interrupt remote player.
+        private Constructable _lastHoveredConstructable = null;
         private readonly IPacketSender packetSender;
-
+        
         public Crafting(IPacketSender packetSender)
         {
+            Instance = this;
             this.packetSender = packetSender;
+            timer.AutoReset = true;
+            timer.Elapsed += new System.Timers.ElapsedEventHandler(OnTimerEvent);
+            timer.Start();
+        }
+
+        //Applies to Fabricator and Workbench
+        public void GhostCrafter_Post_OnHandHover(GameObject gameObject, GUIHand hand)
```

Comment:
Please look at our contributing guidelines and make sure all code is consistent with the rest of the code base:
https://github.com/SubnauticaNitrox/Nitrox/blob/master/CONTRIBUTING.md

Just a few observations in this method:
* No snake case for method naming; this is only acceptable for patch class names.
* No need to fully qualify static calls - use imports to reduce noise.
* No need to pragma debug log statements as we can disable debug log level for releases.  This will reduce the amount of noise.
* Avoid "_" in local names as this is not consistent with the rest of the code base.

Reply:
Sorry, for violating the guidelines. I've been taught naming rules different than used here (but that's quite a while ago now).

- No need to fully qualify > Okay, i will change this.
- Avoid "_" in local names > Okay, i will change this.

To the other two Points, i would like to explain, why i changed the way of naming and usage.

Snake case naming:

For my understanding, the main purpose of this mod is to handle events from the local game-logic and the remote game-logic (server or other players). Events from the local game-engine come from patched methods and can be Pre, Post or Injected via transpiler. The patched method itself is settled in a base-class of the Subnautica-Assembly. Looking at the used method names i can tell at a glance, when this method is called and where it comes from. The first part (class name) and the last part (patched method) are exactly the same as used in Subnautica.

Pragma debug:

That's another thing i've been thaught different. And for me, this way of using dev infos is the best i've ever seen. Why?

I've started with the master from 3 weeks ago and on every log file check i've hundreds of lines from the spawning investigation which are scattered around many files. They relate to investigations of an other Contributor and are not really relevant to me. And i can't easily swith them off. I must go to each line and uncomment it. 
Placing log-statements for investigation and bug fixing is necessary. To remove them to have clean code is not necessary. If any Issue is opened related to this logic part, the bugfixing developer starts all over again and puts in again many lines of logging and removes it after solving the issue. It's work that doesn't need to be done. But the log-output should only conentrate on the investigated theme(s).  
Using normal log statements will have the need to make them filterable (not only debug or trace on/off), which means to extend the logger-class with some filtering-logic and maintain the filters that can be used. Filtering is especially necessary when having OnHover logging or similar.
Not all dev-logging is intended to be in the release logging. Even not if debug log is activated at the customer side. 
When using pragma in the right way:
- Every other developer can take the classes and build them and nothing comes out to logging by default. He has to explicitly activate the pragmas before build. In normal use, no need to manage and apply any filters first for blending out the log-statements of other functions. 
- Additional defines can be easily set by everyone without touching logger-classes and they have no effect on someone others logging needs (the only thing to keep in mind is, that the stated pragmas in the .csproj file must not be included in a pull request). 
- No need to concern about which information is logged in a normal debug or release build, because these lines don't get even compiled in these builds
- Normal logging statements are always executed, even if they are filtered later. This may have impacts e.g. if every second hover action or runtime methods are logged.

Reply:
For the two contention points:

Snake Case - Only patch classes should be snake case to easily differentiate the class and method that are being patched.  Having gamelogic understand that it is being invoked by a patch is inappropriate coupling.  GameLogic classes, such as this Crafter class, should have no knowledge of being invoked via pre/post as they may also be invoked by a packet processor or some other process.  Ideally these methods be made more generic like `RegisterCrafterInUse()`

Pragma - My professional background is not in c# so I had to do a little research.  I'm actually not able to find sources advocating this style of logging even after expanding my search to c++.  To me the trade-off is quite clear, having 3 lines per log statement(two of which are not indented) will be unreadable if you were to have several log statements in a single method.  If desired, the same behavior could be achieved via an enumeration with similar benefits excluding the mostly negotiable compile time benefits.  However, coming from a language without pragmas, and an environment where we log almost every detail, I've never really seen the need but wouldn't protest if we did that.